implement adyen payment reporting

zubair-arbi opened the pull request:
SOL-1880
@hasnain-naveed @ibrahimahmed443 @douglashall 
- Implement `Adyen` daily payment reporting
- Add tests config section for `Adyen`
- Add unit tests for `Adyen` reporting
- Update payment acceptance tests for `Adyen` reports
### How to run this locally
- Start your analytics devstack http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/installation/analytics/start_analytics.html
- Add settings for `adyen` to your `edx-analytics-pipeline/config/devstack.cfg`

```
[adyen]
company_account = TestAccount
merchant_accounts_list = EdXORG
username = report
password = ************
url = https://ca-test.adyen.com/reports/download
interval_start = 2016-08-19
```
- Run the remote command against your `analytics devstack` from your `edx-analytics-pipeline` for a specific day for example to get report for `2016-08-17` run following command:

``` shell
remote-task --vagrant-path ~/Workspace/analytics-edx/analyticstack/ --remote-name devstack --override-config ${PWD}/config/devstack.cfg --wheel-url $WHEEL_URL --branch origin/zub/sol-1880-implement-adyen-transactions-reporting --wait DailyProcessFromAdyenTask --company-account EdXAccount --merchant-account EdXORG --output-root hdfs://localhost:9000/edx-analytics-pipeline/output/adyen/ --run-date=2016-08-17 --local-scheduler
```
- SSH your `analytics` devstack
- Switch to `hadoop` user

```
sudo su hadoop
```
- Verify that there is respective report available by running this command:

```
hadoop@precise64:/home/vagrant$ hadoop fs -ls /edx-analytics-pipeline/output/adyen/
Found 1 items
drwxr-xr-x   - hadoop supergroup          0 2016-08-31 10:32 /edx-analytics-pipeline/output/adyen/payments
```

zubair-arbi commented:
@douglashall @HassanJaveed84 please give initial review for this PR.

zubair-arbi commented:
@brianhw @mulby please take a look at this PR and guide me how to write tests for it.
Since `adyen` provides `tsv` files so we are using `tsv` directly instead of fetching `csv` files firstand then converting them to `tsv` and example `tsv` file is attached on the related JIRA ticket [SOL-1880](https://openedx.atlassian.net/browse/SOL-1880).

HassanJaveed84 commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,160 @@
+        return url
+
+
+class IntervalPullFromAdyenTask(DailyPullFromAdyenTask, WarehouseMixin, luigi.WrapperTask):
```
Should this be PullFromAdyenTaskMixin instead of DailyPullFromAdyenTask ?

zubair-arbi replied:
Yes. Let me change it.

HassanJaveed84 commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,160 @@
+        args = {
+            'company_account': self.company_account,
+            'merchant_account': self.merchant_account,
+            'output_root': self.warehouse_path,
```
I think you can just pass in self.output_root 
```'output_root': self.output_root,```
here as it would default to the warehouse_path if None.

mulby commented:
@zubair-arbi - First, thanks for building this! We are definitely on the right track here. Here's my initial round of feedback:

1. You can see examples of similar tests for the paypal report pull here: https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/reports/tests/test_paypal.py#L534
2. You will also need to update our acceptance tests to include some fake data here: https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/tests/acceptance/test_financial_reports.py#L30
3. As you develop this, please be *very* careful not to accidentally upload the password for this account. I recommend copying the devstack.cfg file somewhere outside of the repo say "cp config/devstack.cfg /var/tmp/adyen.cfg", modifying it to include the real password and then running with `LUIGI_CONFIG_PATH=/var/tmp/adyen.cfg launch-task ...`
4. Can you point me to a spec for the format of the TSV that we download from adyen? If it contains certain classes of protected information about the user then we will not want to persist the file to S3, and instead just extract the information we need and save that result. We need the final output to look like this:  https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/reports/cybersource.py#L150.

zubair-arbi commented:
4) @mulby I have attached a `tsv` file `received_payments_report_2016_08_17.tsv` is on the related [JIRA ticket SOL-1880](https://openedx.atlassian.net/browse/SOL-1880). You can download it and check the fields.

mulby commented:
@zubair-arbi awesome, thanks for the sample. I've responded on the JIRA ticket.

zubair-arbi commented:
@mulby @brianhw I have updated this PR to read/use selected fields from actual `adyen` payments report by using the column names. Please review.

zubair-arbi commented:
@mulby @brianhw @douglashall I have updated this PR by adding a separate task for processing the daily pulled adyen reports and verified that the new task `DailyProcessFromAdyenTask` works locally. Please review.

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+        section_name = 'adyen'
+        self.username = config.get(section_name, 'username')
+        self.password = config.get(section_name, 'password')
+        self.url = config.get(section_name, 'url')
```
I recognize that interval_start is needed in a different class than username, password, and url, but I would move all four out of this mixin and into their respective classes.   I think this mixin should just define the common luigi parameters that need to be passed through.

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+    remote Adyen account with a report user and generates a CSV file.
+
+    Subscribe Adyen daily reports "Payment Accounting Report" with CSV format.
+    https://docs.adyen.com/developers/reporting-manual#reportdownload
```
Is there a more specific link that points directly to the payment accounting report?

zubair-arbi replied:
@brianhw since this class `DailyPullFromAdyenTask` is related to just fetching the subscribed reports so this [report download manual url](https://docs.adyen.com/developers/reporting-manual#reportdownload) from the `reporting-manual` makes more sense here which describes how you can subscribe and automate the report downloading.

For the report structure of `Payment Accounting Report` I have added the [paymentaccountingreportstructure url](https://docs.adyen.com/developers/reporting-manual#paymentaccountingreportstructure) in the method `run` of class `DailyProcessFromAdyenTask` where we process the downloaded report.

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+
+
+class DailyPullFromAdyenTask(PullFromAdyenTaskMixin, luigi.Task):
+    """ A task that reads out adyen reports "Payment Accounting Report" from a
```
nit: capitalize Adyen

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+    Subscribe Adyen daily reports "Payment Accounting Report" with CSV format.
+    https://docs.adyen.com/developers/reporting-manual#reportdownload
+
+    Pulls are made for only a single day. It allows runs to performed
```
nit:  "to performed" => "to be performed".

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+            ), self.password
+        )
+        print "report_url =======> ", self.report_url
+        print "auth =======> ", auth
```
I assume in production we will want the print statements to be removed.

zubair-arbi replied:
This is just for local testing purpose and will be removed with our finalized commit.

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+                    # Output most of the fields from the original source file.
+                    # https://docs.adyen.com/developers/reporting-manual#paymentaccountingreportstructure
+
+                    # Read only settled (sale) or refunded (refund) transaction records
```
nit: add trailing period.

brianhw commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,253 @@
+        if self.output_root is None:
+            self.output_root = self.warehouse_path
+        if self.interval is None:
+            self.interval = date_interval.Custom(self.interval_start, self.interval_end)
```
Since interval_start is only needed here, I think this is where it should be read from configuration.

zubair-arbi replied:
Done.

mulby commented on requirements/pre.txt:
```diff
@@ -1 +1 @@
-pip==1.5.6             # MIT // AN-4322
+pip==8.1.2             # MIT
```
This change seems unrelated to this PR and kind of risky. Is it needed?

zubair-arbi replied:
Previously I was getting issues in installing packages `cffi` and `cryptography` with errors like:
```
build/temp.macosx-10.11-intel-2.7/_openssl.c:433:10: fatal error: 'openssl/opensslv.h' file not found
#include <openssl/opensslv.h>
         ^
1 error generated.
error: command 'cc' failed with exit status 1
``` 
and
```
File "/Users/zubairafzal/Workspace/analytics-edx/edx-analytics-pipeline/venv/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 76, in wrapper
    ffi.def_extern(name=name, **kwargs)(func)
AttributeError: 'CompiledFFI' object has no attribute 'def_extern'
```
So I updated `pip` too while updating some other packages. 

This update is not trivial but I was really surprised to see that we are using such and old version `1.5.6` while the latest version is `8.1.2` with lots of changes and improvements. My suggestion would be to update our packages on consistent basis. On my local machine I had no issues using this version but I you think that is not necessary or we can have a separate PR then I can do as you suggest.

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,261 @@
+
+                    result = [
+                        # date (ISO8601 formatted transaction creation date)
+                        booking_date.date().isoformat(),
```
Updated

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,261 @@
+        Example URL:
+            https://ca-test.adyen.com/reports/download/MerchantAccount/EdXORG/payments_accounting_report_2016_08_17.csv
+        """
+        url = '{base_url}/MerchantAccount/{merchant_account}/{report_name}_{report_date}.{report_format}'.format(
```
According to your suggestion I have update the PR to use UTC everywhere.

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,261 @@
+        print "report_url =======> ", self.report_url
+        print "auth =======> ", auth
+
+        response = requests.get(self.report_url, auth=auth)
```
I have added retry logic, please review.

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,261 @@
+    def __init__(self, *args, **kwargs):
+        super(DailyPullFromAdyenTask, self).__init__(*args, **kwargs)
+
+        self.username = self.config.get(self.config_section_name, 'username')
```
Updated

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,261 @@
+    """
+    # Date to fetch Adyen report.
+    run_date = luigi.DateParameter(
+        default=date.today(),
```
Done

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,261 @@
+        super(PullFromAdyenTaskMixin, self).__init__(*args, **kwargs)
+
+        self.config = get_config()
+        self.config_section_name = 'adyen'
```
I have now moved this as a file level setting `ADYEN_CONFIG_SECTION`.

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,305 @@
+TRANSACTION_TYPE_MAP = {
+    'Settled': 'sale',
+    'Refunded': 'refund'
+}
```
added comment

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,305 @@
+    )
+    username = luigi.Parameter(
+        default_from_config={'section': ADYEN_CONFIG_SECTION, 'name': 'username'},
+        description='Username for the Adyen account, for which data is being gathered for.',
```
done

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,305 @@
+
+        response = fetch_report_with_retry()
+        return response
+
```
removed extra line

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,305 @@
+
+    @property
+    def report_url(self):
+        """Generate the url to download a report from a Adyen account.
```
changed

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,305 @@
+            base_url=self.url,
+            merchant_account=self.merchant_account,
+            report_name=self.REPORT_NAME,
+            report_date=self.run_date.strftime('%Y_%m_%d'),
```
Since this is the making of external `adyen` report url which needs a specific format of date so cannot modify it.

zubair-arbi commented:
@brianhw @mulby @douglashall I have addressed your feedback and also added tests for different scenarios. Please review.

brianhw commented on requirements/test.txt:
```diff
@@ -7,5 +7,6 @@ nose-ignore-docstring==0.2
 nose==1.3.4
 pep8==1.6.0
 pylint==1.5.0
+responses==0.5.1
```
This has a number of dependencies, and it would be good to freeze them as well.   These include requests, and cookies.  The version for requests should be the same as in default.txt, otherwise a newer version may be installed.  It appears this also wasn't done for six, Jinja2, and python-dateutil -- not related to your change, but if you could also freeze those to their default.txt values, that would be great.

zubair-arbi replied:
I have pinned tests dependencies `cookies==2.2.1`, `Jinja2==2.7.3`, `python-dateutil==2.2`, `requests==2.7.0` and `six==1.6.1` w.r.t to our default requirements.

brianhw commented on edx/analytics/tasks/reports/tests/test_adyen.py:
```diff
@@ -0,0 +1,153 @@
+
+    DEFAULT_DATE = '2016-08-17'
+
+    @responses.activate
```
I'm not really familiar with either, but is there a particular advantage to using responses rather than httpretty as was used in test_paypal.py?  It seems a little odd to use two different packages if they both accomplish roughly the same kind of testing of payment provider APIs.

zubair-arbi replied:
I have to use `responses` in order to mock the secure http request (fetching `adyen` reports) since `httpretty` was unable to mock those calls and I couldn't find a way to achieve this other than monkey patching the `httpretty`. I know this seems redundant to use two different libraries for same purpose but I wanted to write tests close to our prod settings. Suggestions are welcome.

mulby replied:
We just test with "http://" urls elsewhere in the code. Could you do that here to avoid pulling in the new dependency?

brianhw commented on edx/analytics/tasks/reports/tests/test_adyen.py:
```diff
@@ -0,0 +1,153 @@
+        self.output_target = FakeTarget()
+        self.task.output = MagicMock(return_value=self.output_target)
+        dummy_adyen_report_path = os.path.join(TEST_DATA_DIR, 'adyen/2016-08/adyen_TestMerchant_20160817.csv')
+        self.dummy_adyen_report = open(dummy_adyen_report_path, 'r').read()
```
This is the first case of having an external data fixture in our unit tests.   Looking in our other tests here in reports, all the others inline the data as a multiline string.

zubair-arbi replied:
I added external data fixture to increase readability (since adding multiple records with different record types can take significant space and also considering individual records to be too long) and improve the unit tests updates in future easily by separating our externally dependent data from our internally processed content.

brianhw commented on edx/analytics/tasks/reports/tests/test_adyen.py:
```diff
@@ -0,0 +1,153 @@
+        requests.codes.request_timeout,         # HTTP Status Code 408
+        requests.codes.too_many_requests,       # HTTP Status Code 429
+        requests.codes.service_unavailable,     # HTTP Status Code 503
+        requests.codes.gateway_timeout          # HTTP Status Code 504
```
nit:  I've no idea if this would work, but can you use *DEFAULT_RETRY_STATUS_CODES here?  Just curious....

zubair-arbi replied:
Yes, this works 👍

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,309 @@
+import requests
+
+from edx.analytics.tasks.url import get_target_from_url
+from edx.analytics.tasks.url import url_path_join
```
updated

zubair-arbi commented on edx/analytics/tasks/reports/adyen.py:
```diff
@@ -0,0 +1,309 @@
+        self.remove_output_on_overwrite()
+
+        # Fetch the report
+        response = self.fetch_report(timeout_seconds=self.REPORT_FETCH_TIMEOUT_SECONDS)
```
I have update the method `fetch_report` to take a required argument (previously optional) `timeout_seconds` to clear the confusion. This will help us in testing the task `DailyPullFromAdyenTask` with custom `timeout_seconds`.

brianhw commented:
LGTM.  👍 after squashing.

mulby commented on edx/analytics/tasks/reports/tests/test_adyen.py:
```diff
@@ -0,0 +1,146 @@
+        output_target_value = self.output_target.value.strip()
+        self.assertEqual(output_target_value, self.dummy_adyen_report)
+
+    @patch('edx.analytics.tasks.reports.adyen.DailyPullFromAdyenTask.REPORT_FETCH_TIMEOUT_SECONDS', 0.5)
```
We try *very* hard to keep our unit tests running very quickly, we currently are able to run ~1700 tests in 9 seconds. I would strongly prefer we don't introduce a 500 ms delay into one of our tests.

I would either not test this logic, or do something like I did here: https://github.com/edx/edx-analytics-pipeline/blob/master/edx/analytics/tasks/util/tests/test_edx_api_client.py#L99

Note that it patches `time.sleep()` and `datetime.utcnow()`

zubair-arbi replied:
@mulby with your suggestion we would have to use timeout of value atleast `0.03` seconds so that there are http calls for adyen reports greater than 1. So now I have removed that test check and replaced timeout with `0.001` seconds which should be fine with our test time target.

zubair-arbi commented:
@mulby @brianhw @douglashall @HassanJaveed84  I have squashed and rebased my PR with master and it's ready for final review. There are some unit tests failing related to `CourseEnrollmentTask` with error `UnknownParameterException` for `output_root` which seems unrelated to my changes.

I have couple of questions too:

1. 
how we add prod config settings for `adyen`?
2. 
would merging this PR now would have any negative affect? since we haven't integrated `Adyen` in out system.

zubair-arbi commented:
@brianhw @mulby The PR is ready for final review. Both unit tests and acceptance tests are passing now.

brianhw commented:
Prod config settings are stored in a separate private repo that we use for storing secure information.  We would need to add Adyen sections to this, to at least define 'merchant_accounts_list' as an empty list.  When we wanted to actually start pulling information from Adyen, we would add entries to 'merchant_accounts_list' and 'company_account'.

Those changes should probably be done in a branch of the secure repo, and then we would run payment-pulling (or all finance) with a different warehouse_path to keep the results separate from production.

mulby commented:
@zubair-arbi - this looks good now, but we need to update our production configuration before merging this to master so that it doesn't break production.

Can you create an analytics ticket that specifies the procedure we can use to obtain the proper reporting credentials for Adyen? We can then follow that procedure, update our production config and merge this to master.

zubair-arbi commented:
@mulby thanks for follow up. I have created an [analytics ticket AN-8051](https://openedx.atlassian.net/browse/AN-8051) for providing procedure to get production configuration settings. The steps aren't defined yet, I have asked @douglashall for help on describing this procedure.

mulby commented:
After some discussion on AN-8051, we've decided to close this until the project is re-prioritized.
